### DISCUSSION TITLE
Shor's Algorithm Implementation - Is it Possible in cudaQuantum ?

### DISCUSSION THREAD

**Discussion by LaibaJamil52** [NVIDIA/cuda-quantum#1552] [Q&A]
I have been trying to implement Shor's Algorithm in cudaQuantum but unable to find a way to implement **modular exponentiation**. Came across Cirq's implementation of Shor's, but it utilizes a built-in class. Is it even possible in cudaQuantum?

Other circuit designs (proposed in research papers) for modular exponentiation are very complex. 

Also can we build a **custom gate** in cudaQuantum?

**Comment by essaathar** [comment]
@bmhowe23, could you please look into the modular exponentiation gate issue here? I've been facing it too.

**Comment by AdityaKoranga** [comment]
Hi any updates on this? I also want to test this.

**Comment by boschmitt** [comment]
[Here](https://nvidia.github.io/cuda-quantum/latest/applications/python/shors.html) is an example of implementing Shor's algorithm using CUDA-Q.

> Also can we build a custom gate in cudaQuantum?

There is a way to build custom unitary gates with CUDA-Q. However, it requires defining the gate using its unitary matrix, see [here](https://nvidia.github.io/cuda-quantum/latest/api/default_ops.html#user-defined-custom-operations).
